### ISSUE TITLE
実装: 競りcontract margin専用NNで平均と標準偏差を同時予測する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#380]
Related: #202, #375, #377, #378
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #378 では multi-head model により contract margin 予測が一定のstate-conditioned signalを持つことを確認した。

- contract margin MAE: `3.074`
- RMSE: `3.832`
- Pearson: `0.532`
- sign accuracy: `94.53%`
- offline pairwise ranking: `66.56%`

一方、現行margin headは各 `(state, action)` に対して平均的なmarginを1値で予測しているだけで、結果の不確実性 / ばらつきを表現しない。

競りAIへ本格利用する場合、例えば同じ予測平均margin `+2` でも、

- ほぼ毎回 `0〜+4` に収まるaction
- `-8〜+12` のように大きくブレるaction

では意味が異なる。

このIssueでは分布全体予測には進まず、まず **contract margin専用NN** を切り出し、各actionについて

- 予測平均 `μ(s,a)`
- 予測標準偏差 `σ(s,a)`

を同時に学習する。

## 目的

競りcompact278 observationから、29 bidding actionsそれぞれについて contract margin の平均と標準偏差を予測する専用modelを実装・評価する。

概念:

```text
compact278
  -> dedicated margin MLP
     -> 29 x mean
     -> 29 x uncertainty parameter
```

初版ではGaussian approximationを用いる heteroscedastic regression を候補とする。

例:

```text
margin ~ Normal(mu(s,a), sigma(s,a)^2)
```

ただし margin が離散・bounded であることは認識し、Gaussian分布全体を最終仕様とみなさない。まず平均+標準偏差という低次元の不確実性表現が有用かを検証する。

## 重要: このsigmaが表すもの

`σ(s,a)` はモデル重みの不確実性そのものではなく、基本的には **入力として観測できないhidden informationやoutcome variationを含む条件付きoutcome dispersion** を表すものとして扱う。

現行compact278は完全情報stateではないため、同様の観測/state-actionでもhidden hand等により最終marginが変わり得る。

必要ならIssueコメントで以下を区別する。

- aleatoric / conditional outcome uncertainty
- epistemic / model uncertainty

初版では前者のみを対象とし、deep ensemble / MC dropout等のepistemic uncertaintyは非目標とする。

## Dataset

まず Issue #377/#378 で使用したfull poolを再利用する。

`/tmp/napoleon-issue377-qdataset-22000x1`

- source states: 22,000
- samples: 122,765
- RuleBased / Conservative 50:50 opponents
- schema v2

fixed validationも #377/#378 と同じstateKey集合を再利用する。

validation stateKey hash:

`9ff43c57867712f5352467b3d56e397046b02c3c11b22a579ab0298a9777d57a`

training / validation split:

- train: 20,000 states
- validation: 2,000 states

NoContract / AllPass sampleはmargin regression lossからmaskする。

## Model

multi-head shared modelではなく、今回は margin専用NN とする。

初版候補:

```text
compact278 -> MLP [512,512,256,256]
          -> mean head: 29
          -> log-variance/log-std head: 29
```

PPO / existing multi-head checkpointから初期化せずfresh initializationを基本とする。

architecture idは明確に分ける。

例:

`bidding-margin-heteroscedastic-mlp-v1`

## Action semantics

現行29 action mappingをそのまま使う。

- 0 = PASS
- 1..28 = target 13..19 x S/H/D/C

各sampleでは `forcedActionIndex` に対応するmean/stdだけをloss対象にする。

teacherが無い他actionへlossを掛けない。

PASS後に最終契約が存在する場合、そのfinal contract marginをPASS actionのteacherとして扱う現行dataset semanticsを確認する。意味が不適切なら、PASS marginの扱いを別途整理してIssueへ報告する。

## Loss

まず heteroscedastic Gaussian negative log likelihood を実装候補とする。

概念:

```text
loss = 0.5 * (log(sigma^2) + (y - mu)^2 / sigma^2)
```

数値安定性のため、networkは直接sigmaを出さず、例えば `logVariance` または `rawScale -> softplus + eps` を使う。

最低限:

- sigma > 0 を保証
- variance collapse / explosion防止
- NaN / inf防止
- 必要ならlog-variance clamp

を実装する。

比較baselineとして、現行 #378 の deterministic margin regression も同じvalidationで評価する。

## Mean予測評価

平均 `mu` について、#378 margin headと同じ指標を出す。

- MAE
- RMSE
- Pearson
- sign accuracy
- suit別
- target別
- pairwise ranking

不確実性を追加したことで平均予測が著しく悪化していないことを確認する。

## Sigma評価: 最重要

単にsigmaを出せたでは不十分。

予測sigmaが実際の誤差/不確実性と対応しているか評価する。

最低限:

1. `|actual margin - predicted mu|` と predicted sigma の関係
2. sigma low/mid/high bucketごとの MAE/RMSE
3. predicted sigmaが大きいsampleほど実誤差が大きいか
4. standardized residual:

```text
z = (actual - mu) / sigma
```

のmean/std/quantile
5. 予測区間coverage

例えばGaussian approximationとして:

- `mu ± 1 sigma` にactualが入る割合
- `mu ± 2 sigma` にactualが入る割合

を出す。

理想的Gaussianなら目安は約68% / 95%だが、marginは非Gaussianの可能性があるので、これを絶対基準にはせず calibration診断として使う。

## Calibration / sharpness

標準偏差が全部同じ定数を出しても役に立たない。

最低限:

- sigma mean/std/min/max
- sigma percentile
- action別sigma
- suit別sigma
- target別sigma
- hand-strength別sigma

を出す。

また、単純baselineとして:

- global residual std一定
- action-index residual std
- target/suit bucket residual std

と比較し、NNのstate-conditioned sigmaが有益かを見る。

## Rankingへの利用

runtime接続はまだしない。

offlineで、平均と標準偏差を使ったrisk-aware scoreを比較する。

最低限:

```text
score_lambda = mu - lambda * sigma
```

について、固定の少数lambdaを比較する。

例:

- lambda = 0.0
- 0.25
- 0.5
- 1.0

大量tuningはしない。

評価:

- best-action hit
- top3
- pairwise ranking
- strongest suit match
- predicted best target distribution
- high-risk action選択率

ただし「best teacher」を何で定義するか明示する。

まずはactual contract marginをteacherとしたoffline rankingで比較する。

## Success probabilityとの関係

#378 contract-success headとの関係も診断する。

例えばGaussian approximationから:

```text
P(margin >= 0)
```

を `mu` と `sigma` から算出できる。

このderived success probabilityと #378 success head を比較する。

最低限:

- ROC-AUC
- PR-AUC
- calibration
- pairwise ranking

を比較し、margin mean+sigmaだけでsuccess headを代替できるか、それともsuccess headを別に残す価値があるか結論を出す。

## Training

#378と比較可能な20k/2k fixed validationを使う。

推奨初期条件:

- hidden dims `[512,512,256,256]`
- lr `1e-3`
- batch size `128`
- regression target standardization:あり/なしを既存実装に合わせて合理的に選択
- validation-based early stopping
- patience 10程度
- CPU/CUDA available device

大規模hyperparameter sweepはしない。

必要ならNLL安定化のためlrのみ軽く調整してよい。

## Baselines

最低限比較:

### Mean
- #378 contract margin head
- global mean
- target mean
- action-index mean
- suit x target mean

### Sigma
- global residual std
- action-index residual std
- suit x target residual std

### Success probability
- #378 explicit success head
- Gaussian derived `P(margin>=0)`

## 判定したいこと

1. Margin専用NNにすると #378 shared multi-head marginよりmean予測が改善するか
2. sigmaはstate/action依存の実誤差と関係するか
3. sigma calibrationは実用レベルか
4. `mu - lambda*sigma` でaction rankingが改善するか
5. Gaussian derived success probabilityはexplicit success headに近づけるか
6. mean+sigmaを競りAIのrisk-aware scoreに使う価値があるか
7. それでも弱いなら、distribution predictionへ進む必要があるか

## Success criteria

有望と判断する目安:

- mean Pearsonが #378 `0.532` を大きく壊さない、できれば改善
- predicted sigmaとabsolute residualに明確な正の関係
- low-sigma bucketのMAEがhigh-sigma bucketより明確に小さい
- 1σ/2σ coverageが極端に崩れていない
- constant-std baselineよりNLL/calibrationで改善
- risk-aware rankingが `lambda=0` より改善する領域がある

## Failure criteria

以下ならsigma headは実用価値が薄いと判断する。

- sigmaがほぼ定数
- sigmaとactual errorに関係がない
- calibrationが大きく崩れる
- mean予測を悪化させる
- risk-aware rankingが一貫して悪化する

その場合、分布全体予測へ無理に進まず、mean margin + explicit success headの方を優先する。

## Tests

最低限:

1. dedicated margin model output shape
2. selected-action-only loss
3. sigma positivity
4. Gaussian NLL fixture
5. extreme log-variance stability
6. NoContract mask
7. stateKey split leakageなし
8. deterministic training smoke
9. checkpoint save/load
10. calibration metric fixture
11. interval coverage fixture
12. risk-aware ranking fixture
13. Gaussian derived success probability fixture
14. ONNX exportする場合PyTorch parity

## Git / PR運用

最新 `feature/issue-202-non-playing-ai` からIssue専用branchを作成する。

PR baseは必ず `feature/issue-202-non-playing-ai`。
main向けPRは作らない。

実装・tests・20k experiment後、Codex review最大4回（review後修正最大3回）。blocking/actionable指摘がなくなればReadyにして親branchへmergeする。

mainにはmergeしない。

## 完了時に必ず報告

- dataset / fixed validation identity
- architecture
- mean/sigma parameterization
- NLL定義
- best epoch
- mean MAE/RMSE/Pearson/sign accuracy
- #378 margin headとの比較
- sigma distribution
- sigma vs absolute residual関係
- low/mid/high sigma bucket MAE
- 1σ/2σ coverage
- standardized residual diagnostics
- constant/bucket std baseline比較
- lambdaごとのoffline ranking
- Gaussian derived success ROC-AUC/PR-AUC
- explicit success headとの比較
- mean+sigma方式が有望か
- runtime bidding scoreへ進めるか
- tests
- PR / merge commit / review回数
- mainへmergeしていないこと
